File: Week_08/52.n-queens-ii.java
/*
 * @lc app=leetcode.cn id=52 lang=java
 *
 * [52] N皇后 II
 *
 * https://leetcode-cn.com/problems/n-queens-ii/description/
 *
 * algorithms
 * Hard (80.07%)
 * Total Accepted:    49.7K
 * Total Submissions: 60.5K
 * Testcase Example:  '4'
 *
 * n 皇后问题研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。
 * 
 * 
 * 
 * 上图为 8 皇后问题的一种解法。
 * 
 * 给定一个整数 n，返回 n 皇后不同的解决方案的数量。
 * 
 * 示例:
 * 
 * 输入: 4
 * 输出: 2
 * 解释: 4 皇后问题存在如下两个不同的解法。
 * [
 * [".Q..",  // 解法 1
 * "...Q",
 * "Q...",
 * "..Q."],
 * 
 * ["..Q.",  // 解法 2
 * "Q...",
 * "...Q",
 * ".Q.."]
 * ]
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 皇后，是国际象棋中的棋子，意味着国王的妻子。皇后只做一件事，那就是“吃子”。当她遇见可以吃的棋子时，就迅速冲上去吃掉棋子。当然，她横、竖、斜都可走一或
 * N-1 步，可进可退。（引用自 百度百科 - 皇后 ）
 * 
 * 
 */
class Solution {
    private int full;
    private int total = 0;
    private void dfs(int queens, int sw, int se) {
        if (queens == full) {
            total++;
            return;
        }
        int row = (~(queens | sw | se)) & full;
        while (row != 0) {
            int pos = row & (-row);
            row -= pos;
            dfs(queens | pos, (sw | pos) << 1, (se | pos) >> 1);
        }
    }
    public int totalNQueens(int n) {
        full = (1 << n) - 1;
        dfs(0,0,0);
        return total;
    }
}

File: Week_08/52.n-queens-ii.py
#
# @lc app=leetcode.cn id=52 lang=python3
#
# [52] N皇后 II
#
# https://leetcode-cn.com/problems/n-queens-ii/description/
#
# algorithms
# Hard (80.07%)
# Total Accepted:    49.7K
# Total Submissions: 60.5K
# Testcase Example:  '4'
#
# n 皇后问题研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。
# 
# 
# 
# 上图为 8 皇后问题的一种解法。
# 
# 给定一个整数 n，返回 n 皇后不同的解决方案的数量。
# 
# 示例:
# 
# 输入: 4
# 输出: 2
# 解释: 4 皇后问题存在如下两个不同的解法。
# [
# [".Q..",  // 解法 1
# "...Q",
# "Q...",
# "..Q."],
# 
# ["..Q.",  // 解法 2
# "Q...",
# "...Q",
# ".Q.."]
# ]
# 
# 
# 
# 
# 提示：
# 
# 
# 皇后，是国际象棋中的棋子，意味着国王的妻子。皇后只做一件事，那就是“吃子”。当她遇见可以吃的棋子时，就迅速冲上去吃掉棋子。当然，她横、竖、斜都可走一或
# N-1 步，可进可退。（引用自 百度百科 - 皇后 ）
# 
# 
#
class Solution:
    def totalNQueens(self, n: int) -> int:
        count, full = 0, (1 << n) - 1
        def dfs(queens, sw, ne):
            nonlocal count
            if queens == full:
                count += 1
                return
            pos = ~(queens | sw | ne) & full
            while pos != 0:
                last = pos & (-pos)
                pos -= last
                dfs(queens | last, (sw | last) << 1, (ne | last) >> 1)
        dfs(0, 0, 0)
        return count
